controlplane/kubeadm/internal/workload_cluster_coredns.go

```diff
@@ -19,35 +19,65 @@ package internal
 import (
 	"context"
 	"fmt"
+	"reflect"
 	"strings"
 
 	"github.com/blang/semver"
-	"sigs.k8s.io/cluster-api/util/version"
-
 	"github.com/coredns/corefile-migration/migration"
 	"github.com/pkg/errors"
 	appsv1 "k8s.io/api/apps/v1"
 	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/util/retry"
 	bootstrapv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1alpha4"
 	controlplanev1 "sigs.k8s.io/cluster-api/controlplane/kubeadm/api/v1alpha4"
 	containerutil "sigs.k8s.io/cluster-api/util/container"
 	"sigs.k8s.io/cluster-api/util/patch"
+	"sigs.k8s.io/cluster-api/util/version"
 	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
 )
 
 const (
-	corefileKey       = "Corefile"
-	corefileBackupKey = "Corefile-backup"
-	coreDNSKey        = "coredns"
-	coreDNSVolumeKey  = "config-volume"
+	corefileKey            = "Corefile"
+	corefileBackupKey      = "Corefile-backup"
+	coreDNSKey             = "coredns"
+	coreDNSVolumeKey       = "config-volume"
+	coreDNSClusterRoleName = "system:coredns"
 
 	kubernetesImageRepository = "k8s.gcr.io"
 	oldCoreDNSImageName       = "coredns"
 	coreDNSImageName          = "coredns/coredns"
 )
 
+var (
+	// Source: https://github.com/kubernetes/kubernetes/blob/v1.22.0-beta.1/cmd/kubeadm/app/phases/addons/dns/manifests.go#L178-L207
+	coreDNSClusterRole = &rbacv1.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      coreDNSClusterRoleName,
+			Namespace: metav1.NamespaceSystem,
+		},
+		Rules: []rbacv1.PolicyRule{
```

Comment:
Can we simplify to only list the rules that need to be updated instead of the whole ClusterRole object? Something like `coreDNS181PolicyRules`